### ISSUE TITLE
Authentication no longer works (v3.0.0)

### ISSUE THREAD

**Issue by fusenuk** [mopidy/mopidy-gmusic#87]
Partially copied from - https://discuss.mopidy.com/t/mopidy-gmusic-not-authenticating/745

The main Mopidy log shows this -

2015-05-26 20:16:36,748 - INFO Starting Mopidy core
2015-05-26 20:16:37,793 - DEBUG ClientLogin()
2015-05-26 20:16:37,843 - INFO Starting new HTTPS connection (1): www.google.com
2015-05-26 20:16:38,510 - DEBUG "POST /accounts/ClientLogin HTTP/1.1" 404 65
2015-05-26 20:16:38,562 - INFO failed to authenticate
2015-05-26 20:16:38,565 - ERROR Failed to login as "MYGoogleUSERNAME"
2015-05-26 20:16:38,572 - INFO Start refreshing Google Music library
2015-05-26 20:16:38,575 - INFO Loaded 0 playlists from Google Music

The gmusicapi log shows this -

!-- begin debug log --!
version: 4.1.0-dev
logging to: /home/mopidy/.cache/gmusicapi/log/gmusicapi.log
2015-05-26 20:16:35,872 - gmusicapi.Mobileclient1 (shared:70) [INFO]: initialized
2015-05-26 20:16:35,880 - gmusicapi.Mobileclient1 (shared:93) [INFO]: logged out
2015-05-26 20:16:37,793 - gmusicapi.Mobileclient1 (shared:204) [DEBUG]: ClientLogin()
2015-05-26 20:16:38,562 - gmusicapi.Mobileclient1 (mobileclient:41) [INFO]: failed to authenticate

I previously had the 'Access for less secure apps' - 'Turned Off' but received the emails saying the logins were being blocked and so set it to 'Turned On'.

The main log suggests that the login request returns a HTTP status 404.

If you go directly to www.google.com/accounts/ClientLogin you also will get a 404 with it showing the text link http://developers.google.com/accounts/docs/AuthForInstalledApps

**Comment by kraiz** [comment]
Yes, having exactly the same problems...

**Comment by dEeds83** [comment]
Also having this problem...since yesterday.

**Comment by kraiz** [comment]
Seems like a change on google-side. See https://github.com/simon-weber/Unofficial-Google-Music-API/issues/338

**Comment by Cysioland** [comment]
That related issue is fixed in the develop branch. Can we have it implemented here?

**Comment by kraiz** [comment]
Did i get it right, that the `android_id` in *gmusic-api*'s login method is the *mopidy-gmusic*'s `device_id`?

Before this change the `device_id` was optional and has been kinda "detected" (here: https://github.com/hechtus/mopidy-gmusic/blob/develop/mopidy_gmusic/session.py#L17). Now, with this change, the `device_id` must be defined to login, so it is required.

Can someone have a look at this? Am i right on this? I'm going to make test this later this week and report back for PR.

**Comment by golobodo** [comment]
With gmusicapi 5.0.0-dev, authentication works. But it isn't released yet. Latest version available trough pip is 4.0.0: https://pypi.python.org/pypi/gmusicapi/4.0.0

Workaround:
pip install -U git+https://github.com/simon-weber/Unofficial-Google-Music-API@develop
pip install -U git+https://github.com/hechtus/mopidy-gmusic@develop

Change both files mentioned here:
https://github.com/golobodo/mopidy-gmusic/commit/38148b25039e1be3f458dd5def846339d31014b4

**Comment by kraiz** [comment]
Can confirm, it works with current develop branch of gmusicapi. But you need to define a `device_id` in mopidy config.

**Comment by Apokly** [comment]
Hey guys,

Thanks for help.
I still have the same problem, even after runniung @golobodo commands.

Could you help me ?

**Comment by morgoth** [comment]
I have exactly the same issue, I tried using the recent version of gmusicapi package

```
2015-06-24 16:19:20,877 INFO [22436:MainThread] gmusicapi.Mobileclient1: !-- begin debug log --!
2015-06-24 16:19:20,878 INFO [22436:MainThread] gmusicapi.Mobileclient1: version: 7.0.0-dev
2015-06-24 16:19:20,878 INFO [22436:MainThread] gmusicapi.Mobileclient1: logging to: /var/lib/mopidy/.cache/gmusicapi/log/gmusicapi.log
2015-06-24 16:19:20,879 INFO [22436:MainThread] gmusicapi.Mobileclient1: initialized
2015-06-24 16:19:20,882 INFO [22436:MainThread] gmusicapi.Mobileclient1: logged out
2015-06-24 16:19:20,884 ERROR [22436:GMusicBackend-3] pykka: Unhandled exception in GMusicBackend (urn:uuid:07546160-7857-42ba-9bd6-1be2f3bcdc2a):
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 191, in _actor_loop
    self.on_start()
  File "/usr/local/lib/python2.7/dist-packages/mopidy_gmusic/actor.py", line 48, in on_start
    self.config['gmusic']['deviceid'])
  File "/usr/local/lib/python2.7/dist-packages/mopidy_gmusic/session.py", line 21, in login
    if not self.api.login(username, password):
TypeError: login() takes exactly 4 arguments (3 given)
```

Is there anyone who solved this?

**Comment by Apokly** [comment]
@morgoth I temporally fixed it when I found the discussion https://discuss.mopidy.com/t/mopidy-gmusic-not-authenticating/745/3 .

This topic explains to change both lines: 

- First, fix login call by setting a random devise_id ("test" in example)
`if not self.api.login(username, password, "test"): `

- Then, automaticly return when fetching thumbs up songs
`def get_thumbs_up_songs(self):
    return {}`

Hope to have an official fix soon.

**Comment by morgoth** [comment]
@Apokly Thanks - it worked!
@hechtus can we have some fix in the repo for this?

**Comment by abarreir** [comment]
@morgoth working on it

**Comment by Apokly** [comment]
Great !
It works with **mopidy-gmusic develop branch** and **Unofficial-Google-Music-API master branch**.

Thank you.

**Comment by iOS-Neil** [comment]
Hi.

I'm still unable to authenticate with Google Music on PiMusicBox.  I've tried absolutely everything I can think of to get this working again to no avail.  Are the fixes above still valid(they don't work for me)?  It's getting to the point where I'm having to consider trying to persuade my other half to switch to Spotify - and she will not be happy with that!

If anyone can advise, please, please help me!

Many thanks,
Neil.

**Comment by fusenuk** [comment]
I'm slightly reluctant to reinstall the whole image to test this for you
iOS-Neil as everything is working smoothly for me

What versions of all the scripts do you have? gmusicapi, mopidy-gmusic etc?

What changes to code have you done?

**Comment by iOS-Neil** [comment]
Hi!

Thank you <so> much for the reply - please DON'T reinstall everything for me if it's working, I'm just over the moon that someone has replied and that its working!

Ok, firstly, I can follow instructions etc., but tinkering with code, I'm not much use. I think I've tried so many different things, im not sure where I'm at. I'm more than happy to reformat & reinstall from scratch if you/someone would be kind enough to give me a quick rundown on what to do?

Again, thank you for replying so quickly - I'm ecstatic that at least it's working for someone!

Neil.

**Comment by fusenuk** [comment]
Well I got it working fairly simply, I put a very rough guide in one of the
threads on the official mopidy discuss forum. I'll try and put it into a
simpler step by step guide tomorrow. Although it was quite a few months ago
and so the code may have changed since then.

**Comment by iOS-Neil** [comment]
That would be great, thanks :)

I've tried changing the line to include username, password, deviceid etc., which didn't work. Will follow any advice you can give and will see how it goes. 

Thanks again :)
Neil.

**Comment by fusenuk** [comment]
So the original way I got it working was installing the pi musicbox image
then logging into the OS via ssh and using the following commands

apt-get install python-crypto

wget
https://github.com/simon-weber/Unofficial-Google-Music-API/archive/develop.zip

pip install --upgrade develop.zip

reboot

Then connect via an sFTP program and do the following edits

"backup and then edit
/usr/local/lib/python2.7/dist-packages/mopidy_gmusic/session.py
3a) somewhere around line 21, you'll need to add "test"(or any string in
quotes) so that the line looks like this:

if not self.api.login(username, password, "test"):

3b) further down the file (line 69) change the get_thumbs_up_songs method
to look like this:

def get_thumbs_up_songs(self): return {}

4) reboot the OS"

This worked back on May, the gmusicapi has definitely changed since then so
that may be where your problems are arising

**Comment by morgoth** [comment]
@iOS-Neil One of the issues I had when fighting with it, was incompatible versions of some pip packages. I don't remember which one it was, but I had to downgrade it, to make dependencies fulfilled

**Comment by iOS-Neil** [comment]
Thanks :) I'll reformat the card and try tonight. Couple of questions if I may: Which version of PiMusicBox are you running? And I don't know if this is a dumb question, but is it possible for me to uninstall the version of gmusicapi that it installs and get the same version as you?

**Comment by iOS-Neil** [comment]
@morgoth - thanks for the info. That's about my limits right there, I'd have no idea how to work any of that out :-|

**Comment by iOS-Neil** [comment]
@fusenek - Hi, did everything exactly as you posted.  Reformatted SD card, fresh install of PiMusicBox, followed all your instructions and it didn't work :( Output from /var/log/mopidy/mopidy.log:

2015-08-25 19:39:58,183 - INFO     Starting Mopidy backends: SpotifyTunigoBackend, GMusicBackend, LocalBackend, StreamBackend, YoutubeBackend
2015-08-25 19:39:58,226 - INFO     Mopidy uses Google Music
2015-08-25 19:39:58,279 - INFO     !-- begin debug log --!
2015-08-25 19:39:58,283 - INFO     version: 4.1.0-dev
2015-08-25 19:39:58,296 - INFO     logging to: /home/mopidy/.cache/gmusicapi/log/gmusicapi.log
2015-08-25 19:39:58,299 - INFO     initialized
2015-08-25 19:39:58,309 - INFO     logged out
2015-08-25 19:39:58,319 - ERROR    Unhandled exception in GMusicBackend (urn:uuid:aa561df6-a92a-4e64-9460-87400492fc7b):
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 191, in _actor_loop
    self.on_start()
  File "/usr/local/lib/python2.7/dist-packages/mopidy_gmusic/actor.py", line 48, in on_start
    self.config['gmusic']['deviceid'])
  File "/usr/local/lib/python2.7/dist-packages/mopidy_gmusic/session.py", line 21, in login
    if not self.api.login(username, password, "test"):
TypeError: login() takes exactly 3 arguments (4 given)
2015-08-25 19:39:58,415 - INFO     Audio output set to "alsasink"
2015-08-25 19:39:58,423 - INFO     Loaded 0 local playlists from /var/lib/mopidy/playlists
2015-08-25 19:39:58,478 - INFO     Loaded 0 local tracks using sqlite
2015-08-25 19:39:58,651 - INFO     Starting Mopidy core
2015-08-25 19:39:58,674 - ERROR    Uncaught exception
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mopidy/commands.py", line 273, in run
    core = self.start_core(mixer, backends)
  File "/usr/local/lib/python2.7/dist-packages/mopidy/commands.py", line 355, in start_core
    return Core.start(mixer=mixer, backends=backends).proxy()
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 93, in start
    obj = cls(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/mopidy/core/actor.py", line 41, in __init__
    self.backends = Backends(backends)
  File "/usr/local/lib/python2.7/dist-packages/mopidy/core/actor.py", line 112, in __init__
    has_library = b.has_library().get()
  File "/usr/lib/python2.7/dist-packages/pykka/future.py", line 299, in get
    exec('raise exc_info[0], exc_info[1], exc_info[2]')
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 470, in ask
    self.tell(message)
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 437, in tell
    raise _ActorDeadError('%s not found' % self)
ActorDeadError: GMusicBackend (urn:uuid:aa561df6-a92a-4e64-9460-87400492fc7b) not found
2015-08-25 19:39:58,689 - INFO     Stopping Mopidy frontends
2015-08-25 19:39:58,693 - INFO     Stopping Mopidy core
2015-08-25 19:39:58,696 - INFO     Stopping Mopidy backends
2015-08-25 19:39:58,718 - INFO     Stopping Mopidy audio
2015-08-25 19:39:58,730 - INFO     Stopping Mopidy mixer

I guess it could be the version I'm using as you suggested?

Thanks,
Neil.

**Comment by fusenuk** [comment]
Hey Neil,
 Did you set the 'deviceid' config setting in /boot/config/settings.ini?

**Comment by iOS-Neil** [comment]
Hey :)

<embarrassed> No I didn't! Will try now. Post back shortly...

**Comment by iOS-Neil** [comment]
Tried that and now can't get back into web gui. This is always because of the extra bit on line 21. Soon as I remove "test" it'll work again (but not authenticate). Wen I checked the device id, there's an Android Device ID and a Google Service Framework. I've tried both of these (not sure which one I should use)...

**Comment by iOS-Neil** [comment]
@fusenuk - don't know if this would work, but do you think there's any chance you could email me the files you have in gmusic_mopidy? Maybe if I tried loading those? Or maybe your session.py? I can pm you my email address...I'm just completely at a loss!

Neil.

**Comment by fusenuk** [comment]
Neil, this is why it isn't working - https://github.com/simon-weber/gmusicapi/commit/ec7ba8b13aa5b03bf5d5ece824fa4bbef8c0353c

The changes to the login function were made 2 days after I had got it working, which at the time required the 4 arguments to the function where as the change made on the 2nd of June I linked above remove the device_id part and only take the 3 arguments, which ties in to what you are saying.

try replacing your /usr/local/lib/python2.7/dist-packages/gmusicapi/ folder contents with this - http://anewpla.net/gmusicapi.rar

those are straight from my musicbox server and are the copies of the files I pulled from github. 

I've tried to find the 'working version' from the archive download page of the gmusicapi page but the files don't match up to what I have, so I think I pulled the files down before he edited them and then assigned a new 4.0 or 5.0 version number of it.

**Comment by iOS-Neil** [comment]
Ok, will try it now (and thanks so much for your continued help)...I'm so sorry but I'm a bit unsure how to do the above - what commands should I issue?

**Comment by fusenuk** [comment]
you need to use an SFTP program (https://winscp.net/eng/docs/free_sftp_client_for_windows) and connect to the machine then browse to the folder I mentioned and just copy and overwrite the files

**Comment by iOS-Neil** [comment]
Hey :)

Did that but didn't work. mopidy.log has lots of errors. 

Realistically, I think I've taken more than enough of your time. I'm really very grateful and it's a shame we couldn't resolve it. I guess I'm going to have to hope for a fix in another version of PiMusicBox in the future. Unless you can make an image of your whole install lol!

Thanks again, it's really appreciated. 

Neil.

**Comment by iOS-Neil** [comment]
@fusenuk - can I ask which version of PiMusicBox you're running? I might try installing that and then making the above changes :)

**Comment by markjewiss** [comment]
@fusenuk - I'm seeing the same issues as Neil. I saw your post above linking to GitHub explaining why it wasn't working. I've used the code that was there to modify my installation, but I'm now getting the error that 3 arguments are required and only 2 are supplied - I think because I haven't been able to find the setup.py file and change it. Any idea where that file is and what my next step is?

**Comment by iOS-Neil** [comment]
@markjewiss - Mark, you're getting "THREE arguments are required"? Hmmm...Which version of PiMusicBox? And you've edited mopidy_gmusic/session.py?

**Comment by markjewiss** [comment]
Hi Neil - yep, I am after making the edits, and that's the file. Sorry, realised my mistake in where I've posted - I'm not using PiMusicBox, just Mopidy natively on a Debian on a brand new RPI. I'm going to work out if I can put a PiMusicBox install onto NOOBS on my RPI and boot into that - will give that a go :)

**Comment by iOS-Neil** [comment]
@markjewiss Let me know how it goes :) For now I took out a free 30 day trial with Spotify in the hope I can get Google up and running before it expires. The wife and kids all use Google so I'm not really prepared to make a switch to Spotify permanent - especially as it can only be used on one device at a time.

**Comment by iOS-Neil** [comment]
WOW!!  I've FINALLY got it up and running!  I've sat for about two hours this morning working through all the errors in the logs, updating bits and pieces, at one point I switched on 2 way authentication - it worked but I've since turned it off again and it's still working :)

To be honest, I've played around so much with different bits, I'm not sure how I got it working(!) But I'm perfectly happy to try and create an image of my sd card so others can try it out - if that would work?

Thanks again to @fusenuk for the pointers :)

(An extremely happy) Neil.

**Comment by markjewiss** [comment]
Great news Neil. An image would be great if you're able to do that :)
On 2 Sep 2015 11:40, iOS-Neil <notifications@github.com> wrote:WOW!!  I've FINALLY got it up and running!  I've sat for about two hours this morning working through all the errors in the logs, updating bits and pieces, at one point I switched on 2 way authentication - it worked but I've since turned it off again and it's still working :)

To be honest, I've played around so much with different bits, I'm not sure how I got it working(!) But I'm perfectly happy to try and create an image of my sd card so others can try it out - if that would work?

Thanks again to @fusenuk for the pointers :)

(An extremely happy) Neil.

**Comment by iOS-Neil** [comment]
I'll try and do it today Mark. Just need to figure out how to make an image first, then I'll test it on another card and upload for you :)

**Comment by fusenuk** [comment]
Sorry I have been abroad the last couple of days so haven't been following
this.

Great news on getting it working Neil, I'd watch out making an image that
you aren't including any personal information or account information in
config files and logs etc

**Comment by iOS-Neil** [comment]
Hey @fusenuk - hope you had a nice time :)

I've edited out all account details in the settings.ini - can you think of anywhere else?  I've just made an image but can make a new one if there are any more changes needed...

**Comment by iOS-Neil** [comment]
Anyone any idea how I can send the link privately?  I don't see a PM facility or anything...?

**Comment by markjewiss** [comment]
You could send me an email? mark jewiss (with no space) at me dot com

Just let me know when you've sent it, that doesn't get checked very often. 

Regards,

Mark.
--
Mark Jewiss
+44 7957 280 901
(sent from a mobile device)

**Comment by iOS-Neil** [comment]
Done :)

**Comment by markjewiss** [comment]
Thanks Neil - will let you know how i get on.

**Comment by markjewiss** [comment]
Finally getting to this... :) What is the file that you zipped up? An image file of the whole OS?

**Comment by iOS-Neil** [comment]
Hi Mark...Yeah, it was the whole SD card that I copied. So I'm hoping you should be able to just extract back to a fresh SD card...

**Comment by iOS-Neil** [comment]
Any joy Mark?

**Comment by markjewiss** [comment]
So today I've actually got it working. I realised I didn't have a fresh SD card, and not wanting to run the risk of having to install NOOBS again if it all went wrong (only been playing with the RPI for a few weeks), I started from scratch with a fresh Raspian install.

Just got it up and running now; it's actually straightforward enough, if you download the DEV packages of GMusic API and MopidyGMusic; that's my other difference, I've installed Mopidy directly, then the plugins I wanted. Just got to sort out a decent client (using MPDroid at the moment, not working for me really), which of course is the advantage you have with using MusicBox...

**Comment by kingosticks** [comment]
All the Web clients can be installed with pip, it should be very easy.

**Comment by markjewiss** [comment]
Yes, agreed. I tried one earlier, and after installation I couldn't start mopidy again. Unfortunately I didn't keep the logs fir reference; I went thru the proper uninstall process to no avail. Only way that I could resolve it was a complete rebuild again. 
Anyway, just installed the MusicBox Web client and that's working well. Need to experiment more tomorrow :)
On 9 Sep 2015 22:16, Nick Steel <notifications@github.com> wrote:All the Web clients can be installed with pip, it should be very easy.
